lib/datasets/downloader.rb

```diff
@@ -31,41 +33,59 @@ def download(output_path)
         headers["Range"] = "bytes=#{start}-"
       end
 
-      Net::HTTP.start(@url.hostname,
-                      @url.port,
-                      :use_ssl => (@url.scheme == "https")) do |http|
-        path = @url.path
-        path += "?#{@url.query}" if @url.query
-        request = Net::HTTP::Get.new(path, headers)
+      start_http(@url, headers) do |response|
+        if response.is_a?(Net::HTTPPartialContent)
+          mode = "ab"
+        else
+          start = nil
+          mode = "wb"
+        end
+
+        base_name = @url.path.split("/").last
+        size_current = 0
+        size_max = response.content_length
+        if start
+          size_current += start
+          size_max += start
+        end
+        progress_reporter = ProgressReporter.new(base_name, size_max)
+        partial_output_path.open(mode) do |output|
+          response.read_body do |chunk|
+            size_current += chunk.bytesize
+            progress_reporter.report(size_current)
+            output.write(chunk)
+          end
+        end
+      end
+      FileUtils.mv(partial_output_path, output_path)
+    rescue TooManyRedirects => error
+      last_url = error.message[/\Atoo many redirections: (.+)\z/, 1]
+      raise TooManyRedirects, "too many redirections: #{@url} .. #{last_url}"
+    end
+
+    private def start_http(url, headers, limit = 10, &block)
+      if limit == 0
+        raise TooManyRedirects, "too many redirections: #{url}"
+      end
+      Net::HTTP.start(url.hostname,
+                      url.port,
+                      :use_ssl => (url.scheme == "https")) do |http|
+        path = url.path
+        path += "?#{url.query}" if url.query
+        request = Net::HTTP::Get.new(url.path, headers)
```

Comment:
We should use `path` here instead of `url.path`.
If we use `url.path`, query parameters aren't used.
I've fixed this.